src/test-utils/image-utils/diff-images.js

```diff
@@ -1,60 +1,68 @@
-const getESq = (dY, dI, dQ) => 0.5053 * dY * dY + 0.299 * dI * dI + 0.1957 * dQ * dQ;
-
-// Get blended r/g/b value after applying alpha
-const applyAlpha = (c, a) => 255 + (c - 255) * a / 255;
-
-/**
- * Get dE square at given index from two pixel arrays
- * @param {Uint8ClampedArray} img1 - pixel data of first image
- * @param {Uint8ClampedArray} img2 - pixel data of second image
- * @param {Number} i - pixel index
- */
-function colorDelta(img1, img2, index) {
-  return Math.sqrt(colorDeltaSq(img1, img2, index));
+/* global Buffer */
+import fs from 'fs';
+import {PNG} from 'pngjs';
+import pixelmatch from 'pixelmatch';
+
+export default function diffImages(source1, source2, options = {}) {
+  return Promise.all([parsePNG(source1), parsePNG(source2)])
+    .then(([image1, image2]) => diffPNGs(image1, image2, options));
 }
 
-function colorDeltaSq(img1, img2, index) {
-  const i = index * 4;
-  const a1 = img1[i + 3];
-  const a2 = img2[i + 3];
+function diffPNGs(image1, image2, options) {
+  const {width, height} = image1;
+  if (width !== image2.width || height !== image2.height) {
+    throw new Error('Image sizes do not match');
+  }
 
-  const r1 = applyAlpha(img1[i + 0], a1);
-  const g1 = applyAlpha(img1[i + 1], a1);
-  const b1 = applyAlpha(img1[i + 2], a1);
+  const {
+    threshold = 0.99,
+    createDiffImage = false,
+    tolerance = 0.1,
+    includeAA = false
+  } = options;
 
-  const r2 = applyAlpha(img2[i + 0], a2);
-  const g2 = applyAlpha(img2[i + 1], a2);
-  const b2 = applyAlpha(img2[i + 2], a2);
+  const diffImage = createDiffImage ? new Uint8Array(width * height) : null;
 
-  return getESq(
-    getY(r1, g1, b1) - getY(r2, g2, b2),
-    getI(r1, g1, b1) - getI(r2, g2, b2),
-    getQ(r1, g1, b1) - getQ(r2, g2, b2)
+  const mismatchedPixels = pixelmatch(
+    image1.data,
+    image2.data,
+    diffImage,
+    width,
+    height,
+    {threshold: tolerance, includeAA}
   );
+
+  const match = 1 - mismatchedPixels / (width * height);
+
+  return {
+    match,
+    matchPercentage: `${(match * 100).toFixed(2)}%`,
+    success: match >= threshold,
+    diffImage
+  };
 }
 
-// TODO - expects imagedata structs
-// may need a helper func to accept different arguments types
-export function diffImagePixels(data1, data2, colorDeltaThreshold = DEFAULT_THRESHOLD) {
-  const pixelCount = data1.data.length / 4;
-  const maxDeltaSq = colorDeltaThreshold * colorDeltaThreshold;
-  let badPixels = 0;
-  for (let i = 0; i < pixelCount; i++) {
-    const delta = colorDeltaSq(data1.data, data2.data, i);
-    if (delta > maxDeltaSq) {
-      badPixels++;
-    }
+function parsePNG(source) {
+  const image = new PNG();
+  if (typeof source === 'string') {
+    // url
+    return new Promise((resolve, reject) => {
+      fs.createReadStream(source)
+        .on('error', reject)
+        .pipe(image)
+        .on('parsed', () => resolve(image));
+    });
+  }
+  if (source instanceof Buffer) {
```

Comment:
Side comment: This is a Node `Buffer`? In loaders.gl I have been trying to standardize on `ArrayBuffer`s to simplify for the users, at least I find it annoying to have to write code for two similar memory classes (even for code only intended for running under Node.js). Perhaps another benefit of moving to loaders.gl/images.

Reply:
puppeteer's `screenshot` method returns a `Buffer`. `pixelmatch` accepts both `Buffer` and `Uint8Array`.